### DISCUSSION TITLE
GCR authentication using long lived SA token

### DISCUSSION THREAD

**Discussion by DragosDumitrache** [renovatebot/renovate#19148] [Request Help] [answered]
Hello folks,

I'm currently in the process of setting up renovate to run in a few repositories using the Mend Github App. Those repositories contain, among other things, Dockerfiles that pull images from a private GCR repository. I've created a SA using READ permissions, similar to dependabot, and followed the steps [here](https://docs.renovatebot.com/docker/#google-container-registry-google-artifact-registry)

Where previously it failed to query the docker registry, now I get the following error:
```
DEBUG: Using queue: host=eu.gcr.io, concurrency=10
DEBUG: GET https://eu.gcr.io/v2/<repository>/<org-name>/<docker-image>tags/list?n=10000 = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=403 retryCount=0, duration=422)
DEBUG: Datasource unauthorized
```

The config I have is the following, with the `<encrypted-token>` being the output of `cat prefixed-service-account.json | base64` encrypted via https://app.renovatebot.com/encrypt
```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:base"
  ],
  "hostRules": [
    {
      "matchHost": "eu.gcr.io",
      "authType": "Basic",
      "encrypted": {
        "token": <encrypted-token>
      }
    }
  ]
}
```
I've also tried specifying `hostType` as `docker`, but not luck. Any help would be appreciated 🙏🏼

**Comment by rarkins** [comment, marked as answer]
Can you test that URL with your token and `curl` to see what result you get?

**Reply by DragosDumitrache** [reply]
I get the following:
```
{"child":[],"manifest":{},"name":"<container-name>","tags":[]}⏎                                                                                     
```

**Reply by rarkins** [reply]
What's the exact curl command you used (redacting the secret of course)?

**Reply by rarkins** [reply]
Please run with `curl -v`

**Reply by DragosDumitrache** [reply]
```
curl -v --header "Accept: application/vnd.docker.distribution.manifest.v2+json" --header "Authorization: Bearer <SECRET>" https://gcr.io/v2/<container-repository-name>/<container-name>/tags/list
*   Trying 142.250.27.82:443...
* Connected to gcr.io (142.250.27.82) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*  CAfile: /etc/ssl/cert.pem
*  CApath: none
* (304) (OUT), TLS handshake, Client hello (1):
* (304) (IN), TLS handshake, Server hello (2):
* (304) (IN), TLS handshake, Unknown (8):
* (304) (IN), TLS handshake, Certificate (11):
* (304) (IN), TLS handshake, CERT verify (15):
* (304) (IN), TLS handshake, Finished (20):
* (304) (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / AEAD-CHACHA20-POLY1305-SHA256
* ALPN, server accepted to use h2
* Server certificate:
*  subject: CN=*.gcr.io
*  start date: Nov  2 13:47:29 2022 GMT
*  expire date: Jan 25 13:47:28 2023 GMT
*  subjectAltName: host "gcr.io" matched cert's "gcr.io"
*  issuer: C=US; O=Google Trust Services LLC; CN=GTS CA 1C3
*  SSL certificate verify ok.
* Using HTTP2, server supports multiplexing
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x13b80c400)
> GET /v2/<container-repository-name>/<container-name>/tags/list HTTP/2
> Host: gcr.io
> user-agent: curl/7.79.1
> accept: application/vnd.docker.distribution.manifest.v2+json
> authorization: Bearer <SECRET> 
< HTTP/2 200 
< docker-distribution-api-version: registry/2.0
< content-type: application/json
< date: Tue, 29 Nov 2022 11:16:56 GMT
< server: Docker Registry
< cache-control: private
< x-xss-protection: 0
< x-frame-options: SAMEORIGIN
< alt-svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000,h3-Q050=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"
< accept-ranges: none
< vary: Accept-Encoding
< 
* Connection #0 to host gcr.io left intact
{"child":[],"manifest":{},"name":"<container-repository-name>/<container-name>","tags":[]}⏎
```

**Reply by rarkins** [reply]
You set authType=Basic in your host rule but in your curl example the auth type is Bearer, so I suggest to delete the authType setting in the host rule

**Reply by DragosDumitrache** [reply]
Alright, I've done that, but I still get the same error. Could the encryption mechanism be the issue here?

**Reply by rarkins** [reply]
Can you double check that what you're curling is the exact same URL as Renovate tries? For example apart from the Basic/Bearer different, I noticed now that you also didn't include the `?n=` part from `https://eu.gcr.io/v2/<repository>/<org-name>/<docker-image>tags/list?n=10000`. Although these may make no difference, it's important not to have such mismatches when you're trying to narrow down the problem.

**Reply by DragosDumitrache** [reply]
That gives me a `no matches for wildcard globbing` error

**Reply by rarkins** [reply]
Do you mean that appending `?n=10000` to the URL makes it go from succeeding to failing?

**Reply by DragosDumitrache** [reply]
Yes, so investigating that failure. That's not the issue renovate is reporting, so I'm just wondering whether renovate's error is a red herring hiding this

**Reply by DragosDumitrache** [reply]
Oh, silly me, that was my shell trying to do some expansion. Appending `?n=10000` gave me the same result

**Reply by rarkins** [reply]
What's your latest config you tested with?

**Reply by DragosDumitrache** [reply]
I got it working, I had to actually add the basic authType back in. But the fundamental issue was the SA didn't actually have read permissions to that repository, so, silly user error
